### PR TITLE
Add IEnumerable and IQueryable connections

### PR DESCRIPTION
**Pull request by myty**
This is to address #19 and the performance implications when creating a connection on large lists.  By changing to `EnumerableSliceMetrics`, we should be able to also utilize `IQueryable` to its full potential.

There are a few more things that I'd like to focus on:
- Add an EFCore sample project to highlight its usefulness of `IEnumerable` and `IQueryable` or update the existing sample projects to do so
- The `EnumerableSliceMetrics` class still needs a little more cleanup
- Write some benchmark tests to see what the performance implications or improvement might be between `EnumerableSliceMetrics` and `ArraySliceMetrics`.

### CODE REVIEW

**Comment by myty** [comment]
I ran a benchmark for the `ToConnection` methods on `IEnumerableConnection` and `ArrayConnection`.  Using a 10,000 record list to hold the data, I ran through a few different scenarios with retrieving records from the beginning of the list and near the end of the list.  For both speed and memory performance, there are significant performance improvements.

``` ini
BenchmarkDotNet=v0.13.1, OS=Windows 10.0.19044.1348 (21H2)
Intel Core i7-7660U CPU 2.50GHz (Kaby Lake), 1 CPU, 4 logical and 2 physical cores
.NET SDK=6.0.100
  [Host]     : .NET 5.0.12 (5.0.1221.52207), X64 RyuJIT
  DefaultJob : .NET 5.0.12 (5.0.1221.52207), X64 RyuJIT
```

| Method                    | pageSize | startIndex |          Mean |         Error |        StdDev |        Median |    Ratio |  RatioSD |       Gen 0 | Allocated |
| ------------------------- | -------- | ---------- | ------------: | ------------: | ------------: | ------------: | -------: | -------: | ----------: | --------: |
| **IEnumerableConnection** | **10**   | **0**      |  **6.339 μs** | **0.0949 μs** | **0.0888 μs** |  **6.332 μs** | **1.00** | **0.00** |  **2.1439** |  **4 KB** |
| ArrayConnection           | 10       | 0          |     22.150 μs |     0.4407 μs |     0.3907 μs |     22.239 μs |     3.50 |     0.10 |     39.9780 |     82 KB |
|                           |          |            |               |               |               |               |          |          |             |           |
| **IEnumerableConnection** | **10**   | **9500**   |  **6.604 μs** | **0.1107 μs** | **0.1035 μs** |  **6.601 μs** | **1.00** | **0.00** |  **2.2354** |  **5 KB** |
| ArrayConnection           | 10       | 9500       |    182.516 μs |     1.9431 μs |     1.8175 μs |    182.496 μs |    27.64 |     0.39 |     39.7949 |     83 KB |
|                           |          |            |               |               |               |               |          |          |             |           |
| **IEnumerableConnection** | **25**   | **250**    | **12.342 μs** | **0.2206 μs** | **0.1955 μs** | **12.354 μs** | **1.00** | **0.00** |  **4.1962** |  **9 KB** |
| ArrayConnection           | 25       | 250        |     34.509 μs |     0.6703 μs |     0.6583 μs |     34.630 μs |     2.79 |     0.09 |     41.6260 |     87 KB |
|                           |          |            |               |               |               |               |          |          |             |           |
| **IEnumerableConnection** | **100**  | **9500**   | **41.907 μs** | **0.9670 μs** | **2.6472 μs** | **40.877 μs** | **1.00** | **0.00** | **14.0991** | **29 KB** |
| ArrayConnection           | 100      | 9500       |    216.566 μs |     2.2395 μs |     1.9852 μs |    216.777 μs |     5.19 |     0.31 |     50.7813 |    106 KB |

**Comment by Shane32** [comment]
@myty Let me know when you are ready for a full review of this PR.  I always figured that this repo was pretty pointless if it didn't handle `IQueryable` -- at least for me anyway.  So I'm glad to see you are taking some initiative to make it more useful.  🥇

**Comment by myty** [comment]
> Let me know when you are ready for a full review of this PR. I always figured that this repo was pretty pointless if it didn't handle `IQueryable` -- at least for me anyway. So I'm glad to see you are taking some initiative to make it more useful. 🥇

@Shane32 I think it's probably at a good spot to go ahead with a full review.  

I've been exploring how we can incorporate GraphQL and Relay into our APIs that we've been wrtiting and I really wanted `IQueryable` support.  I started to go down the route of just adding on top within our own internal libraries and realized it could be something that would be beneficial to everyone.  😃

I have some other improvements that I'd like to contribute as well and had to keep backing off in this PR so that it wasn't massive and incorporating stuff that wasn't relevant. I've got one issue out there to improve the relay sample application, but I've got a few more.

**Comment by myty** [comment]
@Shane32 One thing I was debating was to remove the `ArraySliceMetrics` all together since it really isn't needed anymore.  I can do that if you think it's ok to do that at this point.  Of course it's possible someone may be have a dependency at this point so maybe marking obsolete it would a better option.

**Comment by myty** [comment]
@Shane32 Let me know if there's anything else that I need to do for this PR, otherwise I think it's good enough for review.  Thanks!

**Comment by Shane32** [comment]
Thanks for reminding me.  I'll try to take a look at it this weekend.

**Comment by myty** [comment]
@Shane32 @sungam3r I added IQueryable to this.  If nothing else needs address, I think it's good for review.

**Comment by Shane32** [comment]
Sounds good. I'll review

File: src/GraphQL.Relay/GraphQL.Relay.csproj
Comment (Shane32) [inline src/GraphQL.Relay/GraphQL.Relay.csproj:4]: Any reason we cannot target `netstandard2.0` ?  And why the package reference to `System.Linq.Queryable`?

**Comment by Shane32** [comment]
@sungam3r I reviewed this.  My only comment is that we should target .NET Standard 2.0 and eliminate the dependency on System.Linq.Queryable.  Do you want to review this also?  I don't have a need for this package, but I'm fine with this PR to bring it up to date and include `IQueryable` support.

File: src/GraphQL.Relay.Test/GraphQL.Relay.Test.csproj
Comment (sungam3r) [inline src/GraphQL.Relay.Test/GraphQL.Relay.Test.csproj:14]:
```suggestion
```

See Tests.props

**Comment by sungam3r** [comment]
I don't use this either. And I even have no time to look through the code. I started review and will continue tomorrow.

**Comment by myty** [comment]
@Shane32 and @sungam3r Whenever you want to take another review, I've addressed the project files with comments.  👍🏻

**Review by Shane32 — APPROVED** [review]
@sungam3r Do you want to review?

**Comment by Shane32** [comment]
@myty Looks like one of the files needs to be re-saved in VS so it has the right encoding.

**Comment by myty** [comment]
> @myty Looks like one of the files needs to be re-saved in VS so it has the right encoding.

Yeah, I was just looking at that.  Will take a look

**Comment by myty** [comment]
@Shane32 Looks like the CodeQL build check is still failing, but I can't tell if that is expected or not.

**Comment by Shane32** [comment]
> @Shane32 Looks like the CodeQL build check is still failing, but I can't tell if that is expected or not.

I wouldn't worry about it

**Comment by sungam3r** [comment]
Yes, I would like to review that PR again. @myty I fixed CodeQL in #148, update your branch, CodeQL workflow should be green now.

**Comment by sungam3r** [comment]
Also I would like to merge #157 before that to be able to track API changes.

**Comment by myty** [comment]
> Also I would like to merge #157 before that to be able to track API changes.

Sounds good.  Will merge thsi evening.

**Comment by myty** [comment]
> #157 was merged

@sungam3r I've merged master in this branch.

**Comment by sungam3r** [comment]
@myty Update `GraphQL.Relay.approved.txt` file with generated text. Just delete old file, and rename ~.received one with old name. This file tracks API changes so we can see if any breaking changes are made.

**Comment by myty** [comment]
> @myty Update `GraphQL.Relay.approved.txt` file with generated text. Just delete old file, and rename ~.received one with old name. This file tracks API changes so we can see if any breaking changes are made.

@sungam3r I updated the GraphQL.approved.txt file and realized extensions were historically in the Utilities namespace so I moved the `ResolveConnecitonContextExtensions` to the `Utilities` folder and updated the namespace

**Comment by sungam3r** [comment]
So, no breaking changes, OK. Give me 2-3 days to read about relay spec and make review.

**Comment by Shane32** [comment]
@sungam3r This should be reviewed and merged before 0.7.  I already reviewed it a while ago, but you had planned to review it also.

**Comment by sungam3r** [comment]
I still need to dig into relay spec, alas. Let's publish 0.7 to fix #160 . We can publish 0.8 at any time.

**Comment by Shane32** [comment]
@sungam3r I already reviewed this.  It's been sitting here, approved, for over a month.  I'll merge it unless you plan to review this promptly.

File: src/GraphQL.Relay/Types/SliceMetrics.cs
Comment (sungam3r) [inline src/GraphQL.Relay/Types/SliceMetrics.cs:63]: Please either write a descriptive comment or remove empty tags.

File: src/GraphQL.Relay/Types/SliceMetrics.cs
Comment (sungam3r) [inline src/GraphQL.Relay/Types/SliceMetrics.cs:82]: true -> \<see langword="true"/> here and for other places

File: src/GraphQL.Relay/Utilities/ResolveConnectionContextExtensions.cs
Comment (sungam3r) [inline src/GraphQL.Relay/Utilities/ResolveConnectionContextExtensions.cs:13]: Please use cref

File: src/GraphQL.Relay.Test/Utilities/EnumerableExtensionsTests.cs
Comment (sungam3r) [inline src/GraphQL.Relay.Test/Utilities/EnumerableExtensionsTests.cs:5]:
```suggestion
```

**Comment by sungam3r** [comment]
I made a quick review.

File: src/GraphQL.Relay/Types/SliceMetrics.cs
Comment (Shane32) [inline src/GraphQL.Relay/Types/SliceMetrics.cs:18]:
```suggestion
```

File: src/GraphQL.Relay/Types/SliceMetrics.cs
Comment (Shane32) [inline src/GraphQL.Relay/Types/SliceMetrics.cs:42]:
```suggestion
```

**Pull request merged**
